src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Extensibility/ProjectExportProvider.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Extensibility
+{
+    /// <summary>
+    /// MEF component which has methods for consumers to get to project specific MEF exports
+    /// </summary>
+    [Export(typeof(IProjectExportProvider))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class ProjectExportProvider : IProjectExportProvider
+    {
+        private SVsServiceProvider ServiceProvider { get; set; }
+
+        [ImportingConstructor]
+        public ProjectExportProvider(SVsServiceProvider serviceProvider)
+        {
+            ServiceProvider = serviceProvider;
+        }
+
+        /// <summary>
+        /// This function requires going to the UI thread to map the hierarchy to the correct project
+        /// instance. 
+        /// </summary>
+        public T GetExport<T>(IVsHierarchy projectHierarchy) where T : class
+        {
+            if (projectHierarchy == null)
+            {
+                throw new ArgumentNullException(nameof(projectHierarchy));
+            }
+
+            T exportedValue = null;
+
+            ThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
Actually - no, i just confirmed it , its safe to do that since it is done through JoinableTaskFactory which switches contexts correctly. We do that all over the place in our Web project system. I just tried to repo it to double check again - it works fine.

---
In reply to: [74526750](https://github.com/dotnet/roslyn-project-system/pull/349#discussion_r74526750) [](ancestors = 74526750,74469333)